packages/flutter/lib/src/widgets/focus_manager.dart

```diff
@@ -1291,57 +1309,39 @@ class FocusScopeNode extends FocusNode {
   /// The node is notified that it has received the primary focus in a
   /// microtask, so notification may lag the request by up to one frame.
   void autofocus(FocusNode node) {
-    assert(_focusDebug('Node autofocusing: $node'));
-    if (focusedChild == null) {
-      if (node._parent == null) {
-        _reparent(node);
-      }
-      assert(node.ancestors.contains(this), 'Autofocus was requested for a node that is not a descendant of the scope from which it was requested.');
-      node._doRequestFocus(findFirstFocus: true);
+    // Attach the node to the tree first, so in _applyFocusChange if the node
+    // is detached we don't add it back to the tree.
+    if (node._parent == null) {
+      _reparent(node);
     }
+
+    assert(_manager != null);
+    assert(_focusDebug('Autofocus scheduled for $node: scope $this'));
+    _manager?._pendingAutofocuses.add(_Autofocus(scope: this, autofocusNode: node));
```

Comment:
Should remove the node if it was disposed?

Reply:
The `node` and the `scope`? I think they're supposed to be owned by widgets so `FocusManager` shouldn't call dispose on `FocusNode`s.

Reply:
Yes, that's correct.  The owner of the `FocusNode` has the responsibility to dispose of it.

Reply:
I mean the node was disposed before applying the autofocus.

Reply:
Ah I see. But I'm not sure if there's a good way to check if a `FocusNode` is already `dispose`d. It seems fine to keep a `dispose`d `FocusNode` around? In `_applyFocusChange` if either of these 2 nodes are detached from the tree then we throw the entry away anyways.

Reply:
Maybe we can remove the `_Autofocus` from the `_pendingAutofocuses ` on `FocusNode.dispose`.
I test this and it works well, but I am still a little bit worried.

Reply:
I think as long as we don't call notifyListeners or add/remove listeners on disposed nodes we're fine. 

The reason the autofocus thing does not try to deal with individual nodes' lifecycle events is to keep the logic simple. To remove entries when `FocusNode.dispose` is called will be more complicated since we keep the information in the manager. We'd have to look up the node that's currently being disposed from the `_pendingAutofocuses` and remove it. In the case the node has already been detached from the tree it may not do anything since the manager could be null.

Reply:
```dart
 testWidgets('Nested Viewports showOnScreen', (WidgetTester tester) async {
    final List<ScrollController> controllersX = List<ScrollController>.generate(10, (int i) => ScrollController(initialScrollOffset: 400.0));
    final ScrollController controllerY  = ScrollController(initialScrollOffset: 400.0);
    final List<List<Widget>> children = List<List<Widget>>.generate(10, (int y) {
      return List<Widget>.generate(10, (int x) {
        return SizedBox(
          height: 100.0,
          width: 100.0,
          child: Focus(autofocus: true, Text('$x,$y')),
        );
      });
    });

    /// Builds a grid:
    ///
    ///       <- x ->
    ///   0 1 2 3 4 5 6 7 8 9
    /// 0 c c c c c c c c c c
    /// 1 c c c c c c c c c c
    /// 2 c c c c c c c c c c
    /// 3 c c c c c c c c c c  y
    /// 4 c c c c v v c c c c
    /// 5 c c c c v v c c c c
    /// 6 c c c c c c c c c c
    /// 7 c c c c c c c c c c
    /// 8 c c c c c c c c c c
    /// 9 c c c c c c c c c c
    ///
    /// Each c is a 100x100 container, v are containers visible in initial
    /// viewport.
    await tester.pumpWidget(
      Directionality(
        textDirection: TextDirection.ltr,
        child: Center(
          child: SizedBox(
            height: 200.0,
            width: 200.0,
            child: ListView(
              controller: controllerY,
              children: List<Widget>.generate(10, (int y) {
                return SizedBox(
                  height: 100.0,
                  child: ListView(
                    scrollDirection: Axis.horizontal,
                    controller: controllersX[y],
                    children: children[y],
                  ),
                );
              }),
            ),
          ),
        ),
      ),
    );
}
```
The first row will add to the tree and then remove within the first frame.

which node do you think should be the primary focus node ?

Reply:
> The first row will add to the tree and then remove within the first frame(because the initialScrollOffset> 0).

It sounds like with the current implementation something in the first row should be auto-focused, and then it will be unfocused so the scope is gonna get the focus,
and with this patch the second row (or whoever got mounted first and remain attached after the first frame) should be autofocused. 

I'm not sure what the expected behavior is in this case.

Reply:
So right,and I am not sure too.
Additionally, some of disposed nodes added to the pending list and try to request focused.

Reply:
Autofocus is ideally only supposed be set on one widget in a build. Setting it on more than one works (it will set it on the first one added), but is not terribly well defined.  For instance, if you replace the autofocus with `autofocus: x == 0 && y == 0`, then it will autofocus the first entry, even if it isn't initially built.

Reply:
If a node is disposed then it's guaranteed to be detached (reattaching disposed nodes back to the tree shouldn't be allowed). In `_applyFocusChange` we do check if any of the nodes are already detached and discard them.

Reply:
That reminds me we didn't have tests to verify the first come first served behavior. I'll add that in a bit.

Reply:
If we scroll the list ,the primary focus node will be changed. A little strange I think.

Reply:
That's only true if you have `autofocus` set to true on more than one widget, which, as I said above, isn't well defined (mainly because of situations like this where things are built later).

Reply:
Got it.